### PR TITLE
blk: use choose_fd for all filehandle references

### PR DESCRIPTION
**Pull request by wjwithagen**
This is a leading part of the changes to implement bluestore on
FreeBSD. Without this invalid indexing in the descriptor arrays
will occur.

Creates a new enum: `blk_access_mode_t`
to describe `BUFFERED` and `DIRECT `mode access
with `choose_fd()` to get the correct file for the typed access
and adds
a pretty-printer
boolean convertor `blk_access_mode_t::buffermode(bool)`

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
@wjwithagen this does not compile.

**Comment by wjwithagen** [comment]
@tchaikov 
This should do it.

File: src/blk/kernel/KernelDevice.cc [hunk 1]
```diff
@@ -52,6 +52,11 @@ using ceph::make_timespan;
 using ceph::mono_clock;
 using ceph::operator <<;
 
+std::ostream& operator<<(std::ostream& os, const blk_access_mode_t buffered);
```
Comment (tchaikov) [inline src/blk/kernel/KernelDevice.cc:55]: this is confusing. the declaration of a function is located in a .cc file, while its implementation is located in anther .cc file.

File: src/blk/kernel/KernelDevice.cc [hunk 2]
```diff
@@ -87,11 +92,11 @@ KernelDevice::KernelDevice(CephContext* cct, aio_callback_t cb, void *cbpriv, ai
 
 int KernelDevice::_lock()
 {
-  dout(10) << __func__ << " " << fd_directs[WRITE_LIFE_NOT_SET] << dendl;
   // When the block changes, systemd-udevd will open the block,
   // read some information and close it. Then a failure occurs here.
   // So we need to try again here.
   int fd = fd_directs[WRITE_LIFE_NOT_SET];
+  dout(10) << __func__ << " fd=" << fd << dendl;
```
Comment (tchaikov) [inline src/blk/kernel/KernelDevice.cc:99]: could you extract this change into a different commit?

File: src/blk/kernel/KernelDevice.cc [hunk 3]
```diff
@@ -414,10 +419,17 @@ bool KernelDevice::get_thin_utilization(uint64_t *total, uint64_t *avail) const
...
   if (!enable_wrt)
     write_hint = WRITE_LIFE_NOT_SET;
-  return buffered ? fd_buffereds[write_hint] : fd_directs[write_hint];
+#else
+  // Without WRITE_LIFE capabilities, only one file is used.
+  // And rocksdb sets this value also to > 0, so we need to catch this here
+  // instead of trusting rocksdb to set write_hint.
+  write_hint = WRITE_LIFE_NOT_SET;
+#endif
+  return buffermode(buffered) == blk_access_mode_t::BUFFERED
```
Comment (tchaikov) [inline src/blk/kernel/KernelDevice.cc:431]: i don't think the extra layer of indirection helps with the readability.

File: src/blk/kernel/KernelDevice.h [hunk 4]
```diff
@@ -27,6 +27,10 @@
 
 #define RW_IO_MAX (INT_MAX & CEPH_PAGE_MASK)
 
+enum struct blk_access_mode_t {
```
Comment (tchaikov) [inline src/blk/kernel/KernelDevice.h:30]: i am wondering if this enum is used other where than `KernelDevice.cc`, if not, how about moving it into `KernelDevice.cc`?

**Comment by tchaikov** [comment]
also, could you reference the previous choose_fd() PR? so we can at least preserve the review history.

File: src/blk/kernel/KernelDevice.h (hunk 4 above)
Comment (wjwithagen) [reply src/blk/kernel/KernelDevice.h:30]: @tchaikov 
I'm currently only changing the code that I need to get BlueStore ready on FreeBSD, but there is another part in the `blk`tree that perhaps deserves an overhoal:
```
./src/blk/zoned/HMSMRDevice.h:35:  std::vector<int> fd_directs, fd_buffereds;
./src/blk/zoned/HMSMRDevice.cc:61:  fd_buffereds.resize(WRITE_LIFE_MAX, -1);
./src/blk/zoned/HMSMRDevice.cc:145:    fd_buffereds[i] = fd;
./src/blk/zoned/HMSMRDevice.cc:164:    if (fcntl(fd_buffereds[i], F_SET_FILE_RW_HINT, &i) < 0) {
./src/blk/zoned/HMSMRDevice.cc:183:  r = posix_fadvise(fd_buffereds[WRITE_LIFE_NOT_SET], 0, 0, POSIX_FADV_RANDOM);
./src/blk/zoned/HMSMRDevice.cc:221:    BlkDev blkdev_buffered(fd_buffereds[WRITE_LIFE_NOT_SET]);
./src/blk/zoned/HMSMRDevice.cc:277:    if (fd_buffereds[i] >= 0) {
./src/blk/zoned/HMSMRDevice.cc:278:      VOID_TEMP_FAILURE_RETRY(::close(fd_buffereds[i]));
./src/blk/zoned/HMSMRDevice.cc:279:      fd_buffereds[i] = -1;
./src/blk/zoned/HMSMRDevice.cc:312:    assert(fd_buffereds[i] >= 0);
./src/blk/zoned/HMSMRDevice.cc:313:    VOID_TEMP_FAILURE_RETRY(::close(fd_buffereds[i]));
./src/blk/zoned/HMSMRDevice.cc:314:    fd_buffereds[i] = -1;
./src/blk/zoned/HMSMRDevice.cc:355:  int r = ::fstat(fd_buffereds[WRITE_LIFE_NOT_SET], &st);
./src/blk/zoned/HMSMRDevice.cc:362:    BlkDev blkdev{fd_buffereds[WRITE_LIFE_NOT_SET]};
./src/blk/zoned/HMSMRDevice.cc:428:  return buffered ? fd_buffereds[write_hint] : fd_directs[write_hint];
./src/blk/zoned/HMSMRDevice.cc:859:    r = ::sync_file_range(fd_buffereds[WRITE_LIFE_NOT_SET], off, len, SYNC_FILE_RANGE_WRITE|SYNC_FILE_RANGE_WAIT_AFTER|SYNC_FILE_RANGE_WAIT_BEFORE);
./src/blk/zoned/HMSMRDevice.cc:1026:  int r = ::pread(buffered ? fd_buffereds[WRITE_LIFE_NOT_SET] : fd_directs[WRITE_LIFE_NOT_SET],
./src/blk/zoned/HMSMRDevice.cc:1154:      r = ::pread(fd_buffereds[WRITE_LIFE_NOT_SET], t, left, off);
./src/blk/zoned/HMSMRDevice.cc:1208:  int r = posix_fadvise(fd_buffereds[WRITE_LIFE_NOT_SET], off, len, POSIX_FADV_DONTNEED);
```

Very likely the `HMSMRDevice` code was writen/derived from the `KernelDevice` code.
As such it would even warant to migrate a lot of the common needed code up into `src/blk/BlockDevice.*`
But perhaps for now that is a too big a step.

File: src/blk/kernel/KernelDevice.cc (hunk 1 above)
Comment (wjwithagen) [reply src/blk/kernel/KernelDevice.cc:55]: I think the declaration should go where the type is declared since they are closely related.

File: src/blk/kernel/KernelDevice.h (hunk 4 above)
Comment (wjwithagen) [reply src/blk/kernel/KernelDevice.h]: @tchaikov 
Because of this, I did move definitions and functions into `src/blk/BlockDevice.*`

File: src/blk/kernel/KernelDevice.cc [hunk 5]
```diff
@@ -414,9 +414,15 @@ bool KernelDevice::get_thin_utilization(uint64_t *total, uint64_t *avail) const
 
 int KernelDevice::choose_fd(bool buffered, int write_hint) const
 {
-  assert(write_hint >= WRITE_LIFE_NOT_SET && write_hint < WRITE_LIFE_MAX);
+#if defined(F_SET_FILE_RW_HINT)
```
Comment (tchaikov) [inline src/blk/kernel/KernelDevice.cc:417]: @wjwithagen could you please extract this change into a separate commit in this changeset?

**Review by tchaikov — APPROVED** [review]

File: src/blk/kernel/KernelDevice.cc (hunk 5 above)
Comment (wjwithagen) [reply src/blk/kernel/KernelDevice.cc:417]: Fixed

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-07-06_12:30:55-rados-wip-kefu-testing-2021-07-06-1559-1-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-07-07_05:44:54-rados-wip-kefu-testing-2021-07-07-1141-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-07-07_10:41:16-rados-wip-kefu-testing-2021-07-07-1652-distro-basic-smithi/

failures are not related.
